adapter/src/main/java/com/twilio/kudu/sql/rules/KuduSortRule.java

```diff
@@ -83,21 +84,24 @@ public boolean canApply(final RelTraitSet sortTraits, final KuduQuery query, fin
               && sortField.direction != RelFieldCollation.Direction.STRICTLY_ASCENDING)) {
         return false;
       }
-      // the sort columns must be a prefix of the primary key columns
-      if (sortField.getFieldIndex() >= openedTable.getSchema().getPrimaryKeyColumnCount()
-          || sortField.getFieldIndex() != pkColumnIndex) {
-        // This field is not in the primary key columns. If there is a condition lets
-        // see if it is there
-        if (filter.isPresent()) {
-          final RexNode originalCondition = filter.get().getCondition();
-          while (pkColumnIndex < sortField.getFieldIndex()) {
-            final KuduFilterVisitor visitor = new KuduFilterVisitor(pkColumnIndex);
-            final Boolean foundFieldInCondition = originalCondition.accept(visitor);
-            if (foundFieldInCondition.equals(Boolean.FALSE)) {
-              return false;
-            }
-            pkColumnIndex++;
+      if (sortField.getFieldIndex() >= openedTable.getSchema().getPrimaryKeyColumnCount()) {
+        return false;
+      }
+      if (sortField.getFieldIndex() != pkColumnIndex) {
+
+        if (predicates == null) {
+          predicates = mq.getAllPredicates(original);
+          if (predicates == null) {
+            return false;
           }
+        }
+        final KuduFilterVisitor visitor = new KuduFilterVisitor(pkColumnIndex);
+        final boolean primaryKeyIncluded = predicates.pulledUpPredicates.stream().filter(rexNode -> {
+          final Boolean matched = rexNode.accept(visitor);
+          return matched != null && matched;
+        }).allMatch(Strong::isStrong);
```

Comment:
why do we need the `isStrong` check?

Reply:
[`KuduFilterVisitor`](https://github.com/twilio/calcite-kudu/blob/94c5f6d9e181196c03ee34ee72ece9d7226b0bc0/adapter/src/main/java/com/twilio/kudu/sql/rules/KuduSortRule.java#L145-L150) only returns TRUE if the `RexInputRef` for the column is present in the filter condition.

Therefore, it will return true for:
1. `ACCOUNT <> 'XYZ'`
2. `ACCOUNT IS NULL`
3. `ACCOUNT > 'ABC'`

I believe [Strong::isStrong](https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/Strong.java#L121-L145) asserts our requirement that the field is definitely required in the query?

Reading through `Strong` again I have gotten confused. What do you think? We do need an extra assertion, I will write a few tests to see if it works rather than trying to reason about Strong.

Reply:
`visitCall` of `KuduFilterVisitor` also checks to see that the filter has an equals or and AND expression that has an equals, so I thought it would return false for the three cases you pointed out, but I guess we need a test for this.

Reply:
ya, so this logic I have written is wrong. `RelMetadataQuery#getAllPredicates(RelNode)` returns:
[RexTableInputRef](https://github.com/apache/calcite/blob/3d13846a13398a1ba6c1fa84a7d0c0cc543f23d4/core/src/main/java/org/apache/calcite/rex/RexTableInputRef.java#L45), which the visitor does not handle. 

`allMatch` stream method returns true if the stream is empty, which it always is in this PR.

Reply:
This is now fixed: https://github.com/twilio/calcite-kudu/pull/2/commits/f673f3dabeac8425c2373a63a5c1906887424e14